Tutorials (#221)
applied-material-modeling/neml2: doc/tutorials/models/model_parameters_revisited/main.md

@insert-title:tutorials-models-model-parameters-revisited

[TOC]

## Problem description

In a [previous tutorial](#tutorials-models-model-parameters), we briefly explained the usage of model parameters and their retrieval and update. In this tutorial, we will introduce other ways of specifying model parameters in the input file and how they facilitate model composition.

> Consider the following equations describing the stress-strain relation accounting for a thermal eigenstrain:

Reviewer: ```suggestion
Consider the following equations describing an elastic stress-strain relation accounting for a thermal eigenstrain:
```

\f[
  \boldsymbol{\sigma} = f(\boldsymbol{\varepsilon}, T; \alpha, K, G)
\f]
to
\f[
  \boldsymbol{\sigma} = f(\boldsymbol{\varepsilon}, T; \mathcal{P}_\alpha, \mathcal{P}_K, \mathcal{P}_G)
\f]
where \f$ \mathcal{P}_\alpha \f$, \f$ \mathcal{P}_K \f$, and \f$ \mathcal{P}_G \f$ denote the parametrizations of \f$ \alpha \f$, \f$ K \f$, and \f$ G \f$, respectively. In our example, these parametrizations are the abscissa and ordinate of the interpolant.

The composed model automatically reflects such restructuring:
<div class="tabbed">

- <b class="tab-title">C++</b>
  @source:src5
  ```cpp
  #include "neml2/models/Model.h"

  using namespace neml2;

  int
  main()
  {
    auto & model = load_model("input3.i", "eq");
    std::cout << model << std::endl;
  }
  ```
  @endsource

  Output:
  ```
  @attach-output:src5
  ```
- <b class="tab-title">Python</b>
  @source:src6
  ```python
  import neml2

  model = neml2.load_model("input3.i", "eq")
  print(model)
  ```
  @endsource

  Output:
  ```
  @attach-output:src6
  ```

</div>

> When the referenced model has more than one output variables, the variable specification becomes ambiguous. A more precise variable specifier in the form of `<model-name>.<variable-name>` can be used to remove such ambiguity. For example, although unncessary (as [ScalarLinearInterpolation](#scalarlinearinterpolation) defines one and only one output variable), the above example input is equivalent to

Reviewer: ```suggestion
When the referenced model has more than one output variable, the variable specification becomes ambiguous. A more precise variable specifier in the form of `<model-name>.<variable-name>` can be used to remove such ambiguity. For example, although unnecessary (as [ScalarLinearInterpolation](#scalarlinearinterpolation) defines one and only one output variable), the above example input is equivalent to
```

    coefficients = 'K.parameters/K G.parameters/G'
  []
  [eq]
    type = ComposedModel
    models = 'eq1 eq2 eq3'
  []
[]
```

Finally, a special form of variable specifier can be used to effectively transform a model parameter into an **input variable**. For example, the following input file converts the coefficient of thermal expansion into an input variable.
```
[Models]
  [K]
    type = ScalarLinearInterpolation
    argument = 'forces/T'
    abscissa = '300 350 400 450'
    ordinate = '1.4e5 1.35e5 1.32e5 1.25e5'
  []
  [G]
    type = ScalarLinearInterpolation
    argument = 'forces/T'
    abscissa = '300 500'
    ordinate = '7.8e4 7e4'
  []
  [eq1]
    type = ThermalEigenstrain
    reference_temperature = '300'
    CTE = 'forces/alpha'
    eigenstrain = 'forces/Eg'
  []
  [eq2]
    type = SR2LinearCombination
    from_var = 'forces/E forces/Eg'
    to_var = 'forces/Ee'
    coefficients = '1 -1'
  []
  [eq3]
    type = LinearIsotropicElasticity
    strain = 'forces/Ee'
    stress = 'state/S'
    coefficient_types = 'BULK_MODULUS SHEAR_MODULUS'
    coefficients = 'K G'
  []
  [eq]
    type = ComposedModel
    models = 'eq1 eq2 eq3'
  []
[]
```

> Again, NEML2 automatically reflects such change in model structure:

Reviewer: ```suggestion
Again, NEML2 automatically reflects such changes in model structure:
```
